components/common/src/command/package/install.rs

```diff
@@ -502,43 +625,139 @@ impl<'a> InstallTask<'a> {
+            );
+        let latest = match (latest_installed, latest_cached) {
+            (Ok(pkg_install), Err(_)) => pkg_install,
+            (Err(_), Ok(pkg_artifact)) => pkg_artifact,
+            (Ok(pkg_install), Ok(pkg_artifact)) => {
+                if pkg_install.as_ref() > pkg_artifact.as_ref() {
+                    pkg_install
+                } else {
+                    pkg_artifact
+                }
+            }
+            (Err(_), Err(_)) => return Err(Error::PackageNotFound),
+        };
+        debug!("offline mode: winner: {:?}", &latest);
+
+        Ok(latest)
     }
 
-    // TODO (CM): This could return a plain bool IF we could ensure
-    // above that the package identifier is FULLY QUALIFIED
-    fn is_artifact_cached(&self, ident: &PackageIdent) -> Result<bool> {
-        Ok(self.cached_artifact_path(ident)?.is_file())
+    fn latest_installed_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        match PackageInstall::load(ident, Some(self.fs_root_path)) {
+            Ok(pi) => FullyQualifiedPackageIdent::from(pi.ident().clone()),
+            Err(_) => Err(Error::PackageNotFound),
+        }
+    }
+
+    fn latest_cached_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        let filename_glob = {
+            let mut ident = ident.clone();
+            if ident.version.is_none() {
+                ident.version = Some(String::from("_VERSION_"));
+            }
+            if ident.release.is_none() {
+                ident.release = Some(String::from("_RELEASE_"));
+            }
+            match ident.archive_name() {
+                Some(name) => name.replace("_VERSION_", "?*").replace("_RELEASE_", "?*"),
+                None => return Err(Error::PackageNotFound),
+            }
+        };
+        let glob_path = self.artifact_cache_path.join(filename_glob);
+        let glob_path = glob_path.to_string_lossy();
+        debug!("looking for cached artifacts, glob={}", glob_path.as_ref());
+
+        let mut latest: Vec<(PackageIdent, PackageArchive)> = Vec::with_capacity(1);
+        for file in glob::glob(glob_path.as_ref())
+            .expect("glob pattern should compile")
+            .filter_map(StdResult::ok)
+        {
+            let mut artifact = PackageArchive::new(&file);
+            let artifact_ident = artifact.ident().ok();
+            if let None = artifact_ident {
+                continue;
+            }
+            let artifact_ident = artifact_ident.unwrap();
+            if artifact_ident.origin == ident.origin && artifact_ident.name == ident.name {
+                if latest.is_empty() {
+                    latest.push((artifact_ident, artifact));
+                } else {
+                    if artifact_ident > latest[0].0 {
+                        let _ = latest.pop();
+                        latest.push((artifact_ident, artifact));
+                    }
+                }
+            }
+        }
```

Comment:
I'm confused by the use of the `Vec` here. Would an `Option` not be preferable?
```rust
        let mut latest: Option<(PackageIdent, PackageArchive)> = None;
        for …
        {
            let mut artifact = PackageArchive::new(&file);

            if let Ok(artifact_ident) = artifact.ident() {
                if artifact_ident.origin == ident.origin && artifact_ident.name == ident.name {
                    if latest.is_none() || artifact_ident > latest.as_ref().unwrap().0 {
                        latest = Some((artifact_ident, artifact));
                    }
                }
            }
        }
```

Reply:
Haha, yeah I agree! I was getting pretty tired by the time I finished this method up and wasn't totally happy with it. I started with an `Option<(PackageIdent, PackageArchive)>` but either got into code obsession with `None` or an ownership issue. What I really wanted was something you could do in C: allocate the memory for the slot, and keep the "winner" as you go. That's what led me to a memory location of size 1, which was a Vec with stack semantics 😄 

I'll give this a few minutes to rework and see if I can get it clearer!

Reply:
Can you use `Option` with a helping of `get_or_insert` and/or `take`? I think that'd effectively allow you to do what you want, with clearer semantics than the `Vec` approach.

Also, since `PackageArchive` can give you an identifier, do you need to keep both? Would it be enough to just hang on to the `PackageArchive`?

Reply:
Would the code I suggested work?

Reply:
@baumanj Maybe... not sure if the code issues / ownership issues @fnichol ran into would be bypassed by your implementation. At first glance, though, I _think_ it'd work. 

I may have been trying to be too clever with the `get_or_insert`/`take` stuff 😄